atc/worker/placement.go

```diff
@@ -281,25 +281,22 @@ func (strategy limitActiveTasksStrategy) Order(logger lager.Logger, pool Pool, w
 }
 
 func (strategy limitActiveTasksStrategy) Approve(logger lager.Logger, worker db.Worker, spec runtime.ContainerSpec) error {
-	if spec.Type != db.ContainerTypeTask {
+	if spec.Type != db.ContainerTypeTask || strategy.MaxTasks == 0 {
 		return nil
 	}
 
-	activeTasks, err := worker.IncreaseActiveTasks()
+	activeTasks, err := worker.ActiveTasks()
```

Comment:
Can there be the case when there are multiple threads fetch the active tasks at the same time and then both increase active tasks even though it goes past the maximum? For example, max tasks = 1 and there are two threads, both fetch that active tasks are 0 (because they run at the exact same time) and then both increase active tasks so then active tasks = 2.

Reply:
IMHO the potential race of increasing active tasks is trivial as long as the next task cant be scheduled to this particular worker.

From the original issue, it is more concern on the mismatch between task container counts and active tasks count in the row, which should not be affected by the race here.

Reply:
But that would mean when the user sets a max tasks of 1, theres the possibility of it getting to 2,3, or more? I think that might be a problem because then the user will be expecting us to cap the mask tasks at 1 but then in reality there are more than 1 task running.

I guess I'm still uncertain as to where the bug in this original code is that will cause the mismatch between task container counts and active tasks counts. 🤔  Which is why I'm not sure if this will solve the problem? Do you have a theory as to why this code can result in a mismatch of counts?

To be clear, I think there is definitely something that needs to be done to fix this problem. Like maybe we can do a `SELECT FOR UPDATE` in order to lock the `active_tasks` row so that it cannot be updated while it is being selected by one thread within the same transaction? But I also don't know if we should do that because doing `FOR UPDATE` locks have not worked out well for us in the past haha. Whenever I search online, people always say to do a straight update is actually the easiest rather than doing a select and then update. Maybe we could have the update only work if it's count is less than the max tasks? Like
```
UPDATE workers SET active_tasks=active_tasks+1 WHERE name = (worker-name) AND active_tasks < (max-tasks-num) RETURNING active_tasks;`
```
And then we can check if the returned `active_tasks` is nil or not. If it is nil that means it did not update, meaning that it is has hit the max active tasks and we can return the ErrTooManyActiveTasks error.

Reply:
OK I have updated the method